### PR TITLE
replace QSORT with stdlib qsort

### LINKED ISSUE #2424: Server crashes in Record_Get
When I run the following queries
```
CREATE (:L {m:7518529121104228127})
CREATE (:L {m:973534261302138660})
CREATE (:L {m:-3094839435357647266})
CREATE (:L {a:624014191080025169, m:-36713081251936641791616188957439570013717933290926389760})
CREATE (:L {m:5447997858573538122})
CREATE (:L {m:5447997858573538122})
MATCH (n:L) SET n.a = NULL RETURN n.a ORDER BY n.m DESC
```
the server crashes:
```
1:M 21 Jun 2022 17:43:03.613 # Redis 6.2.5 crashed by signal: 11, si_code: 1
1:M 21 Jun 2022 17:43:03.613 # Accessing address: 0x800000118
1:M 21 Jun 2022 17:43:03.613 # Crashed running the instruction at: 0x7fd6b6cde5ad

------ STACK TRACE ------
EIP:
/usr/lib/redis/modules/redisgraph.so(Record_Get+0xd)[0x7fd6b6cde5ad]

Backtrace:
/lib/x86_64-linux-gnu/libpthread.so.0(+0x12730)[0x7fd6b8ccf730]
/usr/lib/redis/modules/redisgraph.so(Record_Get+0xd)[0x7fd6b6cde5ad]
/usr/lib/redis/modules/redisgraph.so(+0x20ccee)[0x7fd6b6ce8cee]
/usr/lib/redis/modules/redisgraph.so(+0x20d77e)[0x7fd6b6ce977e]
/usr/lib/redis/modules/redisgraph.so(ExecutionPlan_Execute+0x48)[0x7fd6b6cdf7f8]
/usr/lib/redis/modules/redisgraph.so(+0x1fe005)[0x7fd6b6cda005]
/usr/lib/redis/modules/redisgraph.so(CommandDispatch+0x407)[0x7fd6b6cdaa37]
redis-server *:6379(RedisModuleCommandDispatcher+0x54)[0x557adf7d7d04]
redis-server *:6379(call+0x9d)[0x557adf750c7d]
redis-server *:6379(execCommand+0x212)[0x557adf7a0282]
redis-server *:6379(call+0x9d)[0x557adf750c7d]
redis-server *:6379(processCommand+0x5b3)[0x557adf7527e3]
redis-server *:6379(processCommandAndResetClient+0x1c)[0x557adf765fdc]
redis-server *:6379(processInputBuffer+0xea)[0x557adf7686ca]
redis-server *:6379(+0xf5298)[0x557adf7f8298]
redis-server *:6379(aeProcessEvents+0x2a1)[0x557adf749ce1]
redis-server *:6379(aeMain+0x1d)[0x557adf749f3d]
redis-server *:6379(main+0x314)[0x557adf7466f4]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xeb)[0x7fd6b8b1e09b]
redis-server *:6379(_start+0x2a)[0x557adf746bda]

------ REGISTERS ------
1:M 21 Jun 2022 17:43:03.615 #
RAX:00000008000000c0 RBX:0000000000002000
RCX:00007fd6b87bb8e8 RDX:0000000000002000
RDI:0000000800000020 RSI:0000000000000002
RBP:0d82b040ad961724 RSP:00007ffea5671c58
R8 :0000000000000000 R9 :00007fd6b8623d20
R10:0000000000000001 R11:0000000000000002
R12:0000000000000000 R13:00007fd6b870d404
R14:0000000000000000 R15:00007fd6b8640780
RIP:00007fd6b6cde5ad EFL:0000000000010206
CSGSFS:002b000000000033
1:M 21 Jun 2022 17:43:03.615 # (00007ffea5671c67) -> 00007fd6b87b9380
1:M 21 Jun 2022 17:43:03.615 # (00007ffea5671c66) -> 0000000000000003
1:M 21 Jun 2022 17:43:03.615 # (00007ffea5671c65) -> 00007ffea5672080
1:M 21 Jun 2022 17:43:03.615 # (00007ffea5671c64) -> 00007fd6b870d40c
1:M 21 Jun 2022 17:43:03.615 # (00007ffea5671c63) -> 00007ffea5671cc0
1:M 21 Jun 2022 17:43:03.615 # (00007ffea5671c62) -> 00007ffea5671cc0
1:M 21 Jun 2022 17:43:03.615 # (00007ffea5671c61) -> 00007fd6b870d434
1:M 21 Jun 2022 17:43:03.615 # (00007ffea5671c60) -> 00007ffea5672080
1:M 21 Jun 2022 17:43:03.615 # (00007ffea5671c5f) -> 0000000000000003
1:M 21 Jun 2022 17:43:03.615 # (00007ffea5671c5e) -> 00007fd6b870d41c
1:M 21 Jun 2022 17:43:03.615 # (00007ffea5671c5d) -> 00007fd6b870d414
1:M 21 Jun 2022 17:43:03.615 # (00007ffea5671c5c) -> 0000000000000001
1:M 21 Jun 2022 17:43:03.615 # (00007ffea5671c5b) -> 00007fd6b87bb8e8
1:M 21 Jun 2022 17:43:03.615 # (00007ffea5671c5a) -> 0000000800000020
1:M 21 Jun 2022 17:43:03.615 # (00007ffea5671c59) -> 00007fd6b87aaa50
1:M 21 Jun 2022 17:43:03.615 # (00007ffea5671c58) -> 00007fd6b6ce8cee

------ INFO OUTPUT ------
# Server
redis_version:6.2.5
redis_git_sha1:00000000
redis_git_dirty:0
redis_build_id:9a0fd17da17df116
redis_mode:standalone
os:Linux 5.18.2-arch1-1 x86_64
arch_bits:64
multiplexing_api:epoll
atomicvar_api:c11-builtin
gcc_version:8.3.0
process_id:1
process_supervised:no
run_id:7c01261ba629a40b23b5e3f291ea85d918ce551c
tcp_port:6379
server_time_usec:1655833383613432
uptime_in_seconds:12
uptime_in_days:0
hz:10
configured_hz:10
lru_clock:11666215
executable:/data/redis-server
config_file:
io_threads_active:0

# Clients
connected_clients:1
cluster_connections:0
maxclients:10000
client_recent_max_input_buffer:8
client_recent_max_output_buffer:0
blocked_clients:0
tracking_clients:0
clients_in_timeout_table:0

# Memory
used_memory:1750448
used_memory_human:1.67M
used_memory_rss:16019456
used_memory_rss_human:15.28M
used_memory_peak:1750448
used_memory_peak_human:1.67M
used_memory_peak_perc:131.60%
used_memory_overhead:845400
used_memory_startup:845328
used_memory_dataset:905048
used_memory_dataset_perc:99.99%
allocator_allocated:1616160
allocator_active:1851392
allocator_resident:5210112
total_system_memory:16715657216
total_system_memory_human:15.57G
used_memory_lua:37888
used_memory_lua_human:37.00K
used_memory_scripts:0
used_memory_scripts_human:0B
number_of_cached_scripts:0
maxmemory:0
maxmemory_human:0B
maxmemory_policy:noeviction
allocator_frag_ratio:1.15
allocator_frag_bytes:235232
allocator_rss_ratio:2.81
allocator_rss_bytes:3358720
rss_overhead_ratio:3.07
rss_overhead_bytes:10809344
mem_fragmentation_ratio:12.63
mem_fragmentation_bytes:14751448
mem_not_counted_for_evict:0
mem_replication_backlog:0
mem_clients_slaves:0
mem_clients_normal:0
mem_aof_buffer:0
mem_allocator:jemalloc-5.1.0
active_defrag_running:0
lazyfree_pending_objects:0
lazyfreed_objects:0

# Persistence
loading:0
current_cow_size:0
current_cow_size_age:0
current_fork_perc:0.00
current_save_keys_processed:0
current_save_keys_total:0
rdb_changes_since_last_save:45
rdb_bgsave_in_progress:0
rdb_last_save_time:1655833371
rdb_last_bgsave_status:ok
rdb_last_bgsave_time_sec:-1
rdb_current_bgsave_time_sec:-1
rdb_last_cow_size:0
aof_enabled:0
aof_rewrite_in_progress:0
aof_rewrite_scheduled:0
aof_last_rewrite_time_sec:-1
aof_current_rewrite_time_sec:-1
aof_last_bgrewrite_status:ok
aof_last_write_status:ok
aof_last_cow_size:0
module_fork_in_progress:0
module_fork_last_cow_size:0

# Stats
total_connections_received:3
total_commands_processed:70
instantaneous_ops_per_sec:0
total_net_input_bytes:2782
total_net_output_bytes:3681
instantaneous_input_kbps:0.00
instantaneous_output_kbps:0.00
rejected_connections:0
sync_full:0
sync_partial_ok:0
sync_partial_err:0
expired_keys:0
expired_stale_perc:0.00
expired_time_cap_reached_count:0
expire_cycle_cpu_milliseconds:0
evicted_keys:0
keyspace_hits:39
keyspace_misses:3
pubsub_channels:0
pubsub_patterns:0
latest_fork_usec:0
total_forks:0
migrate_cached_sockets:0
slave_expires_tracked_keys:0
active_defrag_hits:0
active_defrag_misses:0
active_defrag_key_hits:0
active_defrag_key_misses:0
tracking_total_keys:0
tracking_total_items:0
tracking_total_prefixes:0
unexpected_error_replies:0
total_error_replies:1
dump_payload_sanitizations:0
total_reads_processed:50
total_writes_processed:47
io_threaded_reads_processed:0
io_threaded_writes_processed:0

# Replication
role:master
connected_slaves:0
master_failover_state:no-failover
master_replid:6db65e8973e2ee5cf15464e0277370259f003176
master_replid2:0000000000000000000000000000000000000000
master_repl_offset:0
second_repl_offset:-1
repl_backlog_active:0
repl_backlog_size:1048576
repl_backlog_first_byte_offset:0
repl_backlog_histlen:0

# CPU
used_cpu_sys:0.021129
used_cpu_user:0.025292
used_cpu_sys_children:0.000000
used_cpu_user_children:0.002030
used_cpu_sys_main_thread:0.020097
used_cpu_user_main_thread:0.024057

# Modules
module:name=graph,ver=999999,api=1,filters=0,usedby=[],using=[],options=[]

# Commandstats
cmdstat_exec:calls=23,usec=3020,usec_per_call=131.30,rejected_calls=0,failed_calls=0
cmdstat_graph.QUERY:calls=20,usec=2890,usec_per_call=144.50,rejected_calls=0,failed_calls=0
cmdstat_multi:calls=24,usec=23,usec_per_call=0.96,rejected_calls=0,failed_calls=0
cmdstat_graph.DELETE:calls=3,usec=80,usec_per_call=26.67,rejected_calls=0,failed_calls=1

# Errorstats
errorstat_ERR:count=1

# Cluster
cluster_enabled:0

# Keyspace
db0:keys=1,expires=0,avg_ttl=0

------ CLIENT LIST OUTPUT ------
id=7 addr=172.17.0.1:59182 laddr=172.17.0.2:6379 fd=8 name= age=0 idle=0 flags=x db=0 sub=0 psub=0 multi=1 qbuf=14 qbuf-free=40940 argv-mem=4 obl=4 oll=0 omem=0 tot-mem=61476 events=r cmd=exec user=default redir=-1

------ CURRENT CLIENT INFO ------
id=7 addr=172.17.0.1:59182 laddr=172.17.0.2:6379 fd=8 name= age=0 idle=0 flags=x db=0 sub=0 psub=0 multi=1 qbuf=14 qbuf-free=40940 argv-mem=4 obl=4 oll=0 omem=0 tot-mem=61476 events=r cmd=exec user=default redir=-1
argv[0]: 'GRAPH.QUERY'
argv[1]: 'db'
argv[2]: 'MATCH (n:L) SET n.a = NULL RETURN n.a ORDER BY n.m DESC'
argv[3]: '--COMPACT'
1:M 21 Jun 2022 17:43:03.616 # key 'db' found in DB containing the following object:
1:M 21 Jun 2022 17:43:03.616 # Object type: 5
1:M 21 Jun 2022 17:43:03.616 # Object encoding: 0
1:M 21 Jun 2022 17:43:03.616 # Object refcount: 1

------ MODULES INFO OUTPUT ------
# graph_executing commands
graph_command:GRAPH.QUERY MATCH (n:L) SET n.a = NULL RETURN n.a ORDER BY n.m DESC

------ FAST MEMORY TEST ------
1:M 21 Jun 2022 17:43:03.616 # Bio thread for job type #0 terminated
1:M 21 Jun 2022 17:43:03.616 # Bio thread for job type #1 terminated
1:M 21 Jun 2022 17:43:03.616 # Bio thread for job type #2 terminated
*** Preparing to test memory region 557adf907000 (2277376 bytes)
*** Preparing to test memory region 557ae12b5000 (2478080 bytes)
*** Preparing to test memory region 7fd6a0000000 (135168 bytes)
*** Preparing to test memory region 7fd6a8000000 (135168 bytes)
*** Preparing to test memory region 7fd6ae400000 (8388608 bytes)
*** Preparing to test memory region 7fd6aec00000 (4194304 bytes)
*** Preparing to test memory region 7fd6af04d000 (2621440 bytes)
*** Preparing to test memory region 7fd6af2ce000 (8388608 bytes)
*** Preparing to test memory region 7fd6afacf000 (8388608 bytes)
*** Preparing to test memory region 7fd6b02d0000 (8388608 bytes)
*** Preparing to test memory region 7fd6b0ad1000 (8388608 bytes)
*** Preparing to test memory region 7fd6b12d2000 (8388608 bytes)
*** Preparing to test memory region 7fd6b1ad3000 (8388608 bytes)
*** Preparing to test memory region 7fd6b22d4000 (8388608 bytes)
*** Preparing to test memory region 7fd6b2ad5000 (8388608 bytes)
*** Preparing to test memory region 7fd6b32d6000 (8388608 bytes)
*** Preparing to test memory region 7fd6b3ad7000 (8388608 bytes)
*** Preparing to test memory region 7fd6b42d8000 (8388608 bytes)
*** Preparing to test memory region 7fd6b4ad9000 (8388608 bytes)
*** Preparing to test memory region 7fd6b52da000 (8388608 bytes)
*** Preparing to test memory region 7fd6b5adb000 (8388608 bytes)
*** Preparing to test memory region 7fd6b62dc000 (8388608 bytes)
*** Preparing to test memory region 7fd6b81fd000 (12288 bytes)
*** Preparing to test memory region 7fd6b8200000 (8388608 bytes)
*** Preparing to test memory region 7fd6b8af6000 (16384 bytes)
*** Preparing to test memory region 7fd6b8cb7000 (24576 bytes)
*** Preparing to test memory region 7fd6b8cda000 (16384 bytes)
*** Preparing to test memory region 7fd6b8fc3000 (16384 bytes)
*** Preparing to test memory region 7fd6b91eb000 (8192 bytes)
*** Preparing to test memory region 7fd6b9218000 (4096 bytes)
.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O.O
Fast memory test PASSED, however your memory can still be broken. Please run a memory test for several hours if possible.

------ DUMPING CODE AROUND EIP ------
Symbol: Record_Get (base: 0x7fd6b6cde5a0)
Module: /usr/lib/redis/modules/redisgraph.so (base 0x7fd6b6adc000)
$ xxd -r -p /tmp/dump.hex /tmp/dump.bin
$ objdump --adjust-vma=0x7fd6b6cde5a0 -D -b binary -m i386:x86-64 /tmp/dump.bin
------
1:M 21 Jun 2022 17:43:04.027 # dump of function (hexdump of 141 bytes):
89f0488d048048c1e0044801f88b505883fa01741b85d274574883ec0883fa02743683fa0474194883c408e9e03cfeff488b5018488b4010c30f1f8000000000e84bffffff4883c4084889c7e9ff3cfeff0f1f8000000000e8c3feffff4883c4084889c7e9d73cfeff0f1f8000000000e99b3cfeff66662e0f1f84000000000089f6488d44b60548c1e004c744
Function at 0x7fd6b6cde530 is Record_GetEdge
Function at 0x7fd6b6cde4c0 is Record_GetNode
```

**Redis Version:** 6.2.5
**RedisGraph Version:** Self built from master branch
**Operating System:** 5.18.2 Arch Linux

### PR DESCRIPTION
**Pull request by swilly22**
Resolves #2424
I've switched from our [QSORT](https://github.com/RedisGraph/RedisGraph/blob/master/src/util/qsort.h) dependency implementation to [stdlib's qsort](https://en.wikipedia.org/wiki/Qsort)

On a 1M node graph sorting all of the nodes using:

```
MATCH (n)
WITH n ORDER BY n.v
RETURN count(n)
```
Runtime dropped from 272ms to 140ms a ~%48 speed up on my development machine.

### CODE REVIEW

File: tests/flow/test_distinct.py [hunk 1]
```diff
@@ -76,7 +76,7 @@ def test_distinct_with_order(self):
         result = graph1.query("MATCH (p:PARENT)-[:HAS]->(c:CHILD) RETURN DISTINCT p.name ORDER BY c.name")
         self.env.assertEqual(result.result_set, [['Stevie'], ['Mike'], ['James']])
 
-        result = graph1.query("UNWIND range(0,3) AS a UNWIND range(4,7) AS b RETURN DISTINCT a ORDER BY b")
+        result = graph1.query("UNWIND range(0,3) AS a UNWIND range(4,7) AS b WITH DISTINCT a ORDER BY b RETURN a ORDER BY a DESC")
```
Comment (AviAvni) [inline tests/flow/test_distinct.py:79]: why this change is needed?

**Review by AviAvni — COMMENTED** [review]
nice! need to understand why memcheck crash and one question

File: tests/flow/test_distinct.py (hunk 1 above)
Comment (swilly22) [reply tests/flow/test_distinct.py:79]: The original order by sorted `b` values which are all the same (4), there's no definition for sorting values that are equal, as such no one can make any assumption on the order.

File: src/execution_plan/ops/op_sort.c [hunk 2]
```diff
@@ -141,25 +150,22 @@ static Record SortConsume(OpBase *opBase) {
 	if(!newData) return NULL;
 
 	if(op->buffer) {
-		QSORT(Record, op->buffer, array_len(op->buffer), RECORD_SORT);
+		qsort_r(op->buffer, array_len(op->buffer), sizeof(Record), op,
```
Comment (AviAvni) [inline src/execution_plan/ops/op_sort.c:153]: is qsort_r is correct?

File: src/execution_plan/ops/op_value_hash_join.c [hunk 3]
```diff
@@ -81,69 +93,87 @@ static bool _binarySearchRightmost(uint *idx, Record *array, uint array_len, int
...
 
-	/* Count how many records share the same node.
-	 * reduce search space by truncating left bound */
-	bool found = _binarySearchRightmost(&rightmost_idx, op->cached_records +
-								  leftmost_idx, record_count - leftmost_idx,
-								  op->join_value_rec_idx, v);
+	// count how many records share the same node
+	// reduce search space by truncating left bound
+	bool found = _binarySearchRightmost(&rightmost_idx,
+			op->cached_records + leftmost_idx, record_count - leftmost_idx,
```
Comment (AviAvni) [inline src/execution_plan/ops/op_value_hash_join.c:143]: maybe introduce local var for this calculation?

File: src/execution_plan/ops/op_sort.c [hunk 4]
```diff
@@ -12,56 +12,60 @@
...
+	return 0;
 }
 
-// Compares two heap record nodes.
-static int _heap_elem_compare(const void *A, const void *B, const void *udata) {
-	OpSort *op = (OpSort *)udata;
-	Record aRec = (Record)A;
-	Record bRec = (Record)B;
-	return _record_compare(aRec, bRec, op);
+static int _buffer_elem_cmp
```
Comment (AviAvni) [inline src/execution_plan/ops/op_sort.c:43]: can we get rid of this func?

File: src/util/qsort.h [hunk 5]
```diff
@@ -1,290 +1,62 @@
...
+
+#if defined _SORT_R_BSD
+// BSD (qsort_r) require argument swap
+
+struct sort_r_data {
+	void *arg;
+	int (*compar)(const void *_a, const void *_b, void *_arg);
+};
+
+static inline int sort_r_arg_swap
```
Comment (AviAvni) [inline src/util/qsort.h:30]: the inline here is not really matter

File: src/execution_plan/ops/op_sort.c (hunk 4 above)
Comment (swilly22) [reply src/execution_plan/ops/op_sort.c:43]: We can't, `qsort` passes a pointers to the elements being compared, in this case `Record*` while our heap passes `Record` to its compare function.

File: src/execution_plan/ops/op_sort.h [hunk 6]
```diff
@@ -19,6 +19,7 @@ typedef struct {
 	uint skip;                  // Total number of records to skip
 	uint limit;                 // Total number of records to produce
 	int *directions;            // Array of sort directions(ascending / desending) for each item.
+	uint record_idx;            // index of current record to return
```
Comment (swilly22) [inline src/execution_plan/ops/op_sort.h:22]: @swilly22 address op_update.c TODO line 24
